sklearn/discriminant_analysis.py

```diff
@@ -23,12 +24,19 @@
 from .utils.validation import check_is_fitted
 from .utils.multiclass import check_classification_targets
 from .utils.extmath import softmax
+from .utils._param_validation import StrOptions, Interval
 from .preprocessing import StandardScaler
 
 
 __all__ = ["LinearDiscriminantAnalysis", "QuadraticDiscriminantAnalysis"]
 
 
+@runtime_checkable
+class HasFit(Protocol):
+    def fit(*args, **kwargs):
+        ...
```

Comment:
@jeremiedbb In this instance we need to validate that `covariance_estimator` has a `fit` method (and have a `_covariance` attribute after `fit`) but here's not a great way to do this right now. I suggested this little hack of using a runtime-checkable `Protocol` to at least validate that there is a `fit` method. I'm wondering what your thoughts are on this? Is this something we could have internally in `utils._param_validation`? Is there an alternative for this you'd prefer?

One caveat for this, as pointed out by @thomasjpfan, is that it will not work on `pypy` CI tests run w/ `python 3.7` so we'd need something to catch that, e.g.

```python
if py_version > (3, 7):
    @runtime_checkable
    class HasFit(Protocol):
        def fit(*args, **kwargs):
else:
    # No val on unsupported 3.7
    HasFit = object
```
and then

```python
    _parameter_constraints = {
         ...
        "covariance_estimator": [HasFit, None],
    }
```

Reply:
Hi. Maybe for such cases  we can introduce a new constraint that will basically check all the attributes of the object being passed and if check that the method is callable.

Consider a constraint `required_attr` type that will accept one more parameter that needs to be validated and one for `required_method` type that will also have one more parameter that checks if the argument is callable or not.

Example:

```python
    _parameter_constraints = {
        "covariance_estimator": [required_attr('covariance_estimator'), required_method('fit')],
    }
```

Reply:
These constraints can later be reused for other estimators as well that require specific method or attribute.

Reply:
In this particular instance, the only part we can validate is that it has the `fit` method since the existence of the `_covariance` attribute is not guaranteed until _after_ `fit` is called. But yeah, a new constraint type that essentially tests `callable(getattr(covariance_estimator, 'fit', None))` would also probably work.

Reply:
@Micky774 yeah we cannot check for `covariance_` attribute but definitely a new constraint for `fit` method should be great !

Reply:
A new constraint for "has a fit method" seems appropriate indeed. I don't think there's a need to involve Protocol. I think checking hasattr(fit) in the `is_satisfied_by` method of the constraint should be enough.

Checking the covariance attribute is impossible since it's not set before fitting. But it's fine to ignore that from this validation. It's not currently tested anyway.

Reply:
@jeremiedbb should I work on the new constraint in this PR itself ?

Reply:
I think it's better to implement the new constraint in a separate PR

Reply:
Ok. So Can I work on it on a separate PR ?

Reply:
So for the moment, we can set `"no_validation"` and make a new PR to check the new constraint.